dhis2/app-runtime
# feat: dhis2 connection status [LIBS-315]

Part of [LIBS-315](https://dhis2.atlassian.net/browse/LIBS-315)

Adds a DHIS2 connection status service as part of the offline tools. Coordinates with the offline interface to deduce the status by incidental network requests, then sends pings periodically if there is no other traffic.

Depends on https://github.com/dhis2/app-platform/pull/718; would also benefit from [DHIS2-14531: Create dedicated ping endpoint](https://dhis2.atlassian.net/browse/DHIS2-14531)

### The spec
(This section is the same text as in the Platform PR above)
* We want to detect connection to the DHIS2 server.
* We want to minimize the network and server resources used to do so
* The service worker watches app network traffic and uses the success or failure of responses to determine connectivity status, then send messages to the client to broadcast the updates
    * Only requests to the DHIS2 server (based on the Base URL) are used to determine the connection status
* If there is no app network traffic to the DHIS2 server, the client will send pings
    * The service worker is configured to only handle pings over the network, i.e. not to cache ping requests/responses, so the result observed on the client will reflect the server connection status
    * If the connection status detected by the ping is the same as the previous status, then the time to the next ping will increase. This will continue to create an exponential back-off, which will max at out at a value, for example 5 minutes.
    * If the connection status has changed (either by ping or by app network traffic), the exponential back-off will be reset and the timeout to the next ping will be reduced back to the initial timeout.
    * If the app loses focus, a “standby” procedure is followed to stop pinging when it’s unnecessary
        * Nothing happens initially. If the user refocuses the app before the next ping is scheduled to be sent, the behavior will not change.
        * If the next-scheduled ping would take place while the app is defocused, instead the hook enters a “standby” mode.
        * When the app is refocused, if the hook is in “standby”, a ping will be sent immediately to detect the current connection status, following other rules for back-off
        * If the app is opened while defocused initially, it will also start running in standby and will not ping until refocused.
    * If the app detects an ‘offline’ event from the browser, it will ping immediately to check the connection status, since that should be a low-cost action
    * If the app detects an ‘offline’ event from the browser and is defocused, a special thing happens: it enters standby to send a ping imediately when the app is refocused again
        * If the app is refocused before the next regularly-scheduled ping, the ping timeout will be restarted but not incremented.
        * Otherwise, if the regularly scheduled ping is ready before the app is refocused, the normal standby procedure is followed (e.g. the next timeout can be incremented if the connection status hasn’t changed)

### Notes for this PR
* The new functionality is made up primarily of two parts: 
    * the SmartInterval class, which handles the timeout backoffs, resets, and snoozes
    * the useDhis2ConnectionStatus hook, which integrates the SmartInterval to handle ping timing and the offline interface to listen to updates from the service worker.
* It would be nice to export one `useOnlineStatus` hook which returns ConnectionStatus by default but could return NetworkStatus by passing an `options.useNetworkStatus` parameter, but that turned out to be complicated for several reasons (for example the rule of hooks). After discussing with Hendrik, we decided to leave the existing hook’s name, and export a new `useDhis2ConnectionStatus` hook
    * Under the hood, the `useOnlineStatus` hook has been renamed to `useNetworkStatus` to better reflect its function. Still, it’s exported as `useOnlineStatus` for backward-compatibility
    * The term “online” is a bit ambiguous, but seems to best match the names of the browser APIs used for NetworkStatus
    * The `useDhis2ConnectionStatus` hook then does what it says on the tin
* Smaller decisions are mentioned in inline comments

### To test locally:
* See the [platform PR](https://github.com/dhis2/app-platform/pull/718) for running these together

### To do:
- [x] Remove console logs when done testing
- [x] If [DHIS2-14531: Create dedicated ping endpoint](https://dhis2.atlassian.net/browse/DHIS2-14531) is finished, undo the 'band aid' in the RestAPILink for the current ping endpoint
- [x] Adjust timeout defaults (example: 30 second initial timeout, 1.5x increment factor, 5 min max timeout?)

### Done:
- [x] Create follow-up issues for todos mentioned in the comments


[LIBS-315]: https://dhis2.atlassian.net/browse/LIBS-315?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
